src/test/java/com/google/devtools/build/lib/actions/ActionCacheCheckerTest.java

```diff
@@ -466,15 +986,153 @@ public MiddlemanType getActionType() {
+          return treeArtifactValue.getMetadata();
+        } else {
+          return null;
+        }
+      }
+
+      if (fileMetadata.containsKey(output)) {
+        return fileMetadata.get(output);
+      }
+      return FileArtifactValue.createForTesting(output);
+    }
+
+    @Override
+    public TreeArtifactValue getTreeArtifactValue(SpecialArtifact output) throws IOException {
+      if (treeMetadata.containsKey(output)) {
+        return treeMetadata.get(output);
+      }
+
+      TreeArtifactValue.Builder tree = TreeArtifactValue.newBuilder(output);
+
+      Path treeDir = output.getPath();
+      if (treeDir.exists()) {
+        TreeArtifactValue.visitTree(
+            treeDir,
+            (parentRelativePath, type) -> {
+              if (type == Dirent.Type.DIRECTORY) {
+                return;
+              }
+              Artifact.TreeFileArtifact child =
+                  Artifact.TreeFileArtifact.createTreeOutput(output, parentRelativePath);
+              FileArtifactValue metadata =
+                  FileArtifactValue.createForTesting(treeDir.getRelative(parentRelativePath));
+              tree.putChild(child, metadata);
+            });
+      }
+
+      Artifact.ArchivedTreeArtifact archivedTreeArtifact =
+          Artifact.ArchivedTreeArtifact.create(output, derivedPathPrefix);
+      if (archivedTreeArtifact.getPath().exists()) {
+        tree.setArchivedRepresentation(
+            archivedTreeArtifact,
+            FileArtifactValue.createForTesting(archivedTreeArtifact.getPath()));
+      }
+
+      return tree.build();
     }
 
     @Override
     public void setDigestForVirtualArtifact(Artifact artifact, byte[] digest) {}
   }
+
+  private static class WriteEmptyOutputAction extends NullAction {
+    public WriteEmptyOutputAction() {
+    }
+
+    public WriteEmptyOutputAction(Artifact... outputs) {
+      super(outputs);
+    }
+
+    @Override
+    public ActionResult execute(ActionExecutionContext actionExecutionContext) {
+      for (Artifact output : getOutputs()) {
+        Path path = output.getPath();
+        if (!path.exists()) {
+          try {
+            FileSystemUtils.writeContentAsLatin1(path, "");
+          } catch (IOException e) {
+            throw new IllegalStateException("Failed to create output", e);
+          }
+        }
+      }
+
+      return super.execute(actionExecutionContext);
+    }
+  }
+
+  private static class InjectOutputFileMetadataAction extends NullAction {
+    private final Artifact output;
+    private final FileArtifactValue[] metadata;
+    private int times;
+
+    public InjectOutputFileMetadataAction(Artifact output, FileArtifactValue... metadata) {
+      super(output);
+
+      this.output = output;
+      this.metadata = metadata;
+    }
+
+    @Override
+    public ActionResult execute(ActionExecutionContext actionExecutionContext) {
+      int index = Math.min(times, metadata.length - 1);
+      if (index >= 0) {
```

Comment:
Shouldn't that not happen? If it should, wouldn't it really mean that metadata is empty?

Reply:
Also, should we fail in else?

Reply:
Changed to use a queue